packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,562 @@
+import type {
+  AccountsControllerAccountAddedEvent,
+  AccountsControllerAccountRemovedEvent,
+  AccountsControllerListMultichainAccountsAction,
+} from '@metamask/accounts-controller';
+import {
+  BaseController,
+  type ControllerGetStateAction,
+  type ControllerStateChangeEvent,
+  type RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import { isEvmAccountType } from '@metamask/keyring-api';
+import type {
+  AccountAssetListUpdatedEventPayload,
+  CaipAssetType,
+  CaipAssetTypeOrId,
+} from '@metamask/keyring-api';
+import type { InternalAccount } from '@metamask/keyring-internal-api';
+import { KeyringClient } from '@metamask/keyring-snap-client';
+import type {
+  GetPermissions,
+  PermissionConstraint,
+  SubjectPermissions,
+} from '@metamask/permission-controller';
+import type {
+  GetAllSnaps,
+  HandleSnapRequest,
+} from '@metamask/snaps-controllers';
+import type { FungibleAssetMetadata, Snap, SnapId } from '@metamask/snaps-sdk';
+import { HandlerType } from '@metamask/snaps-utils';
+import {
+  hasProperty,
+  isCaipAssetType,
+  parseCaipAssetType,
+  type CaipChainId,
+} from '@metamask/utils';
+import type { Json, JsonRpcRequest } from '@metamask/utils';
+import type { MutexInterface } from 'async-mutex';
+import { Mutex } from 'async-mutex';
+import { v4 as uuid } from 'uuid';
+
+import { getChainIdsCaveat } from './utils';
+import type { AccountsControllerAccountAssetListUpdatedEvent } from '../../../accounts-controller/src/AccountsController';
+
+const controllerName = 'MultichainAssetsController';
+
+export type MultichainAssetsControllerState = {
+  metadata: {
+    [asset: CaipAssetType]: FungibleAssetMetadata;
+  };
+  allNonEvmTokens: { [account: string]: CaipAssetType[] };
```

Comment:
Also, I think we should prefer using `asset` rather than `token` in the naming of these things as it's commonly used to represent a non-native asset and this controller will also handle the native asset of the chain.

Reply:
Im fine with updating `allNonEvmTokens` to `allNonEvmAssets` 🙏

Reply:
updated here [daaad69](https://github.com/MetaMask/core/pull/5138/commits/daaad69328c964aca96bbcb59d68814d8e378bcf)